### PR TITLE
Ledger response parser

### PR DESCRIPTION
**Pull request by mirgee**
Extracts ledger response parsing capability into the `indy-ledger-response-parser` crate which
* returns types from `indy-data-types` (reexported by `indy-vdr`) and
* defines its own types (taken from `libvdrtools`) for ledger responses.

This crate is further integrated into `IndyVdrLedger` implementation to replace the current provisional constructions.

### CODE REVIEW

File: aries_vcx/src/core/profile/modular_libs_profile.rs [hunk 1]
```diff
@@ -21,10 +22,16 @@ pub struct ModularLibsProfile {
...
 impl ModularLibsProfile {
     pub fn new(wallet: Arc<dyn BaseWallet>, ledger_pool_config: LedgerPoolConfig) -> VcxResult<Self> {
-        let ledger_pool = Arc::new(IndyVdrLedgerPool::new(ledger_pool_config)?);
-        let submitter = Arc::new(IndyVdrSubmitter::new(ledger_pool));
-        let ledger = Arc::new(IndyVdrLedger::new(Arc::clone(&wallet), submitter));
         let anoncreds = Arc::new(IndyCredxAnonCreds::new(Arc::clone(&wallet)));
+        let ledger_pool = Arc::new(IndyVdrLedgerPool::new(ledger_pool_config)?);
+        let request_submitter = Arc::new(IndyVdrSubmitter::new(ledger_pool));
+        let response_parser = Arc::new(ResponseParser::new());
+        let config = IndyVdrLedgerConfig {
```
[the remark this answers is not in the document]
Comment (Patrik-Stas) [reply aries_vcx/src/core/profile/modular_libs_profile.rs:29]: Maybe more related to PR - not sure if `request_submitter`, `response_parser` need to be Arc. I think `IndyVdrLedger` is going to be the only one to use them

File: aries_vcx_core/src/errors/mapping_vdrtools.rs [hunk 2]
```diff
@@ -1,11 +1,23 @@
-use vdrtools::types;
-use vdrtools::types::errors::IndyErrorKind;
+#[cfg(feature = "vdrtools")]
+use vdrtools::types::{
+    errors::{IndyError, IndyErrorKind},
+    ErrorCode,
+};
+
+#[cfg(all(
+    not(feature = "vdrtools"),
```
Comment (Patrik-Stas) [inline aries_vcx_core/src/errors/mapping_vdrtools.rs:8]: wouldn't simply stating this one basically imply 
```
any(feature = "modular_libs", feature = "vdr_proxy_ledger")
```
? Cause if you don't build `vdrtools`, you must be building with something else, no?

I might be missing something

File: aries_vcx_core/src/errors/mod.rs [hunk 3]
```diff
@@ -7,5 +7,5 @@ mod mapping_indyvdr;
 #[cfg(feature = "vdr_proxy_ledger")]
 mod mapping_indyvdr_proxy;
 mod mapping_others;
-#[cfg(feature = "vdrtools")]
+#[cfg(any(feature = "vdrtools", feature = "modular_libs", feature = "vdr_proxy_ledger"))]
```
Comment (Patrik-Stas) [inline aries_vcx_core/src/errors/mod.rs:10]: isn't it redundant? Can there be case that we build without any of these?

File: aries_vcx_core/src/ledger/indy_vdr_ledger.rs [hunk 4]
```diff
@@ -248,89 +260,20 @@ where
...
-    async fn get_schema(&self, schema_id: &str, submitter_did: Option<&str>) -> VcxCoreResult<String> {
-        let _ = submitter_did;
-        // TODO - future - try from cache first
-        // TODO - future - do we need to handle someone submitting a schema request by seq number?
-
-        let id = SchemaId::from_str(schema_id)?;
-
-        let request = self.request_builder()?.build_get_schema_request(None, &id)?;
-
+    async fn get_schema(&self, schema_id: &str, _submitter_did: Option<&str>) -> VcxCoreResult<String> {
```
Comment (Patrik-Stas) [inline aries_vcx_core/src/ledger/indy_vdr_ledger.rs:263]: damn that's so much better, awesome 🥇

File: aries_vcx_core/src/errors/mod.rs (hunk 3 above)
Comment (mirgee) [reply aries_vcx_core/src/errors/mod.rs:10]: It is more to clarify that one of these must be enabled to even build the module (regardless we have a "convention" of always building with one of them).

File: aries_vcx_core/src/errors/mapping_vdrtools.rs (hunk 2 above)
Comment (mirgee) [reply aries_vcx_core/src/errors/mapping_vdrtools.rs:8]: Using your suggestion, if you build with `vdrtools` (default feat. flag) and one of `modular_libs` and `vdr_proxy_ledger`, your suggestion would fail due to duplicate import.

File: aries_vcx/src/core/profile/modular_libs_profile.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (mirgee) [reply aries_vcx/src/core/profile/modular_libs_profile.rs:29]: >  instead of putting things in this config, we could just tweak ndyVdrLedger::new signature and pass it directly. It's not like we have 10 optional things in the "config", all these things are pretty esential.

I put them into config very deliberately. I am planning to add more arguments and I am strongly convinced it is way easier to manage a single config object than many separate arguments.

File: aries_vcx_core/src/ledger/indy_vdr_ledger.rs
Comment (Patrik-Stas) [inline aries_vcx_core/src/ledger/indy_vdr_ledger.rs:1]: Since we support for V0 and V1 responses, I think we should:
- either run tests with both V0 and V1 configurations (should be matter of indy-vdr configuration, what version of ledger protocol we want to use)
- or simply choose to primarily support V1, not test with V0, keep implementation as is, but log warnings anytime V0 response arrives

I suppose indy-vdr would be using V1 rn, but it would be good to make sure

Comment (Patrik-Stas) [inline aries_vcx_core/src/ledger/indy_vdr_ledger.rs:1]: Since we support for V0 and V1 responses, I think we should:
- either run tests with both V0 and V1 configurations (should be matter of indy-vdr configuration, what version of ledger protocol we want to use)
- or simply choose to primarily support V1, not test with V0, keep implementation as is, but log warnings anytime V0 response arrives

I suppose indy-vdr would be using V1 rn, but it would be good to make sure

File: libvdrtools/indy-api-types/Cargo.toml [hunk 5]
```diff
@@ -13,15 +13,15 @@ rust-base58 = ["bs58"]
...
+futures = { version = "0.3", default-features = false, features = ["std"] }
 log = { version = "0.4.17", features = ["std"] }
 libc = "0.2.114"
 openssl = {version = "0.10", optional = true}
 bs58 = {version = "0.4.0", optional = true}
 serde = "1.0.99"
 serde_json = "1.0.40"
 serde_derive = "1.0.99"
-sqlx = { version = "0.5.8", git = "https://github.com/jovfer/sqlx", branch = "feature/json_no_preserve_order_v5", features = [ "sqlite", "json_no_preserve_order" ], optional = true }
+sqlx = { version = "0.5.8", git = "https://github.com/jovfer/sqlx", branch = "feature/json_no_preserve_order_v5", features = [ "sqlite", "json_no_preserve_order", "runtime-tokio-rustls" ], optional = true }
```
Comment (Patrik-Stas) [inline libvdrtools/indy-api-types/Cargo.toml:24]: seems unrelated to pr, why the addition of `runtime-tokio-rustls`?

File: libvdrtools/indy-api-types/Cargo.toml [hunk 6]
```diff
@@ -13,15 +13,15 @@ rust-base58 = ["bs58"]
 
 [dependencies]
 failure = "0.1.8"
-futures = { version = "0.3", default-features = false }
+futures = { version = "0.3", default-features = false, features = ["std"] }
```
Comment (Patrik-Stas) [inline libvdrtools/indy-api-types/Cargo.toml:16]: why needed?

**Review by Patrik-Stas — CHANGES_REQUESTED** [review]
left some comments but nothing big, this is really great improvement

File: libvdrtools/indy-api-types/Cargo.toml (hunk 5 above)
Comment (mirgee) [reply libvdrtools/indy-api-types/Cargo.toml:24]: On the `main` branch, `indy-api-types` and some of its dependents fail to build by themselves because

> one of the features ['runtime-actix-native-tls', 'runtime-async-std-native-tls', 'runtime-tokio-native-tls', 'runtime-actix-rustls', 'runtime-async-std-rustls', 'runtime-tokio-rustls'] must be enabled

The reason why `libvdrtools` or `aries_vcx` don't fail to build is because they add one of these features higher up the dependency chain.

File: libvdrtools/indy-api-types/Cargo.toml (hunk 6 above)
Comment (mirgee) [reply libvdrtools/indy-api-types/Cargo.toml:16]: Same reason as above: `indy-api-types` fails to build without it. Actually, thanks to your comment, I found out the piece of code using the feature seems to have been dead, so I removed it along with the feature.

File: aries_vcx_core/src/ledger/indy_vdr_ledger.rs
Comment (mirgee) [reply aries_vcx_core/src/ledger/indy_vdr_ledger.rs:1]: I am not sure I understand what you mean by "right now", but I think `indy-vdr` can be configured to use either protocol version and we set V1 by default.

**Review by Patrik-Stas — APPROVED** [review]

**Review by bobozaur — CHANGES_REQUESTED** [review]
Looks good, but I think there are some improvements that can be made, especially regarding `parse_message`.

File: indy_ledger_response_parser/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+        )?; // FIXME: Review how we handle not found
+
+        match message {
+            Message::Reject(response) | Message::ReqNACK(response) => Err(err_msg(
+                IndyErrorKind::InvalidTransaction,
+                format!("Transaction has been failed: {:?}", response.reason),
+            )),
+            Message::Reply(reply) => Ok(reply),
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/lib.rs]: I agree that this piece is an ugly workaround. Your suggestion is good in spirit, but it needed a little bit of tweaking to work, as it incorrectly assumes that `type` is a sibling field of `op`, but is in fact part of `Reply` result. To be specific, this is a simplified version of a response:
```
{
  "op": "REPLY",
  "result": {
     ...
    "type": "107"
  }
}
```
(The part about `#[serde(try_from = "MsgWithReplyType")]` on `Message` seems to contradict the rest of your suggestion, where you are proposing to deserialize into `MsgWithTypedReply`, and then convert directly into `Reply`. Moreover, assuming you mean `MsgWithTypedReply`, it would also take generic parameters to account for. But I assume it is no longer relevant.)

File: indy_ledger_response_parser/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+                        format!("Nym not found"),
+                    ))
+                    .and_then(|data| {
+                        serde_json::from_str(&data).map_err(|err| {
+                            IndyError::from_msg(
+                                IndyErrorKind::InvalidState,
+                                format!("Cannot parse GET_NYM response: {}", err),
+                            )
+                        })
+                    })?;
```
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/lib.rs]: Good point about `ok_or_else`, addressed. I don't find extracting into a variable more readable.

File: indy_ledger_response_parser/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+        let (revoc_reg_def_id, revoc_reg) = match reply.result() {
+            GetRevocRegDeltaReplyResult::GetRevocRegDeltaReplyResultV0(res) => (res.revoc_reg_def_id, res.data),
+            GetRevocRegDeltaReplyResult::GetRevocRegDeltaReplyResultV1(res) => {
+                (res.txn.data.revoc_reg_def_id, res.txn.data.value)
+            }
+        };
+
+        let revoc_reg_delta = RevocationRegistryDeltaV1 {
+            value: json!(UrsaRevocationDelta::from_parts(
```
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/lib.rs]: Good point! Fixed.

File: indy_ledger_response_parser/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bobozaur) [reply indy_ledger_response_parser/src/lib.rs]: Aaaahhh, my bad! It was obvious but I missed that. It can still be done in the same fashion though, but it becomes a bit annoying because of the manually specified bounds. I'll close this and start a new convo to make the suggestion.

File: indy_ledger_response_parser/src/domain/response.rs [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+{
+    type Error = IndyError;
+    fn try_from(value: MessageWithTypedReply<'a, T>) -> Result<Self, Self::Error> {
+        match value {
+            MessageWithTypedReply::ReqNACK(r) => Ok(Message::ReqNACK(r)),
+            MessageWithTypedReply::Reply(r) => Ok(Message::Reply(r.result().try_into()?)),
+            MessageWithTypedReply::Reject(r) => Ok(Message::Reject(r)),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/domain/response.rs:129]: It doesn't work, you seem to have reverted the change which actually made it possible to deserialize into `MessageWithTypedReply`.

[the remark this answers is not in the document]
Comment (bobozaur) [reply indy_ledger_response_parser/src/domain/response.rs:129]: Ok lemme actually test. Give me a few minutes.

[the remark this answers is not in the document]
Comment (bobozaur) [reply indy_ledger_response_parser/src/domain/response.rs:129]: Yeah so, if there was any doubt that I'm an idiot, I just provided the confirmation 😅 . I made the exact same mistake that you pointed out before, about the `type` being a field in the `result` and not alongside it.

Modified the suggestion (after testing) to accommodate the field being at the right level in the JSON.

Tested locally with this:
```rust
#[test]
fn test_reply_deserialization() {
    #[derive(Deserialize, Debug)]
    struct Test {
        field: u8,
    }

    impl ReplyType for Test {
        fn get_type<'a>() -> &'a str {
            "107"
        }
    }

    let raw_str = r#"{
        "op": "REPLY",
        "result": {
            "field": 0,
            "type": "107"
        }
      }"#;

    let reply: Message<Test> = serde_json::from_str(raw_str).unwrap();
    println!("{reply:?}");
}
```

File: indy_ledger_response_parser/src/domain/response.rs [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+#[derive(Debug, Deserialize)]
+pub struct ReplyV1<T> {
+    pub data: ReplyDataV1<T>,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct ReplyDataV1<T> {
+    pub result: Vec<ReplyV0<T>>,
```
Comment (bobozaur) [inline indy_ledger_response_parser/src/domain/response.rs:41]: While working on the suggestion to avoid double deserialization, I was wondering why is this a `Vec`, since it looks like in the `result()` method the first element is removed and the others, if any, are lost. If there's only one element to be expected, how about making it an array?
The element can then be extracted with `result.into_iter().next().unwrap()` which is sound since you cannot instantiate an unfilled array.

```suggestion
    pub result: [ReplyV0<T>; 1],
```

**Review by bobozaur — COMMENTED** [review]
Modified the last suggestion for the `Message<T>`, and also wondering about this `Vec` in `ReplyV1`. I'm curious of what your thoughts are on this.

File: indy_ledger_response_parser/src/domain/response.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/domain/response.rs]: Good, the `Reply` version is preserved now! Applied the suggestion.

File: indy_ledger_response_parser/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,244 @@
+#[macro_use]
+extern crate serde;
+
+#[macro_use]
+extern crate serde_json;
+
+mod domain;
+
+use domain::response::MessageWithTypedReply;
```
[the remark this answers is not in the document]
Comment (mirgee) [reply indy_ledger_response_parser/src/lib.rs]: Removed.

File: indy_ledger_response_parser/src/domain/response.rs (hunk 11 above)
Comment (mirgee) [reply indy_ledger_response_parser/src/domain/response.rs:41]: If there is always only one element to be expected, I don't see the point of making it any kind of array in the first place. I don't know why it is a `Vec` - perhaps this is to support some sort of batch transactions in the future...? Then the current implementation of using only the first element would be a temporary workaround, and so I would probably prefer to not accommodate the type to the current use case. But again, as I said, I don't know.

Comment (bobozaur) [reply indy_ledger_response_parser/src/domain/response.rs:41]: Why not deal with the idea of batch transactions if/when it will be a thing? Whether the workaround is temporary or not, I'd say it could at least be a better workaround. But feel free to ignore this and resolve the conversation if it's not important for you.

**Review by bobozaur — DISMISSED** [review]
LGTM

File: indy_ledger_response_parser/src/domain/response.rs (hunk 11 above)
Comment (mirgee) [reply indy_ledger_response_parser/src/domain/response.rs]: Alright, changed the `Vec` to an array.

**Review by bobozaur — APPROVED** [review]
LGTM

**Review by Patrik-Stas — APPROVED** [review]
